### PR TITLE
stage: Skip run-cache if any out has cache: false. 

### LINKED ISSUE #4428: DVC Repro incorrectly saving directory to cache
## Bug Report

UPDATE: Skip to https://github.com/iterative/dvc/issues/4428#issuecomment-778259232

As the title states. I run `dvc repro extract` for my `extract` pipeline stage. This takes two pretty large zip files and extracts them into specified folders. These folders should not be added to the dvc cache, since they can be easily reproduced by extracting the archives, but I declare them as dependencies so that the DAG looks nicer.

My `dvc.yaml` looks like this. The `preprocess` stage should only indicate, that the not cached folders should be used as dependencies in later stages.

```yaml
stages:
  extract:
    cmd: tar -xzvf data/thingy10k/10k_tetmesh.tar.gz -C data/thingy10k/ && tar -xzvf data/thingy10k/10k_surface.tar.gz -C data/thingy10k/
    deps:
    - data/thingy10k/10k_surface.tar.gz
    - data/thingy10k/10k_tetmesh.tar.gz
    outs:
    - data/thingy10k/10k_surface:
        cache: false
    - data/thingy10k/10k_tetmesh:
        cache: false
  preprocess:
    cmd: some preprocessing command
    deps:
    - data/thingy10k/10k_tetmesh
    - data/thingy10k/10k_surface
```

After running `dvc repro extract`, the hashes of all the files are computed and then the files saved to cache. This is exactly the thing that I was trying to prevent with the `cache: false` option. 

I confirmed that the contents of the output folders were indeed added to the cache by using `du -sh .dvc/cache`, which went up by exactly the size of the two folders after running the command.

Interestingly, after running `dvc gc -a` the cache is freed again. Also running `dvc push` (**without** first running `dvc gc -a`) to push the cache to my remote storage also says everything is up to date, which leads me to believe that dvc recognizes that these directories should in fact not be cached.

I have reproduced this in a local git and dvc repo by first adding the two archives using `dvc add` and then running the above mentioned extract stage. The archives can be downloaded from this repo https://github.com/Yixin-Hu/TetWild#dataset under `Output`.

### Please provide information about your setup

**Output of `dvc version`:**

```console
$ dvc version
DVC version: 1.6.0 (pip)
---------------------------------
Platform: Python 3.7.6 on Linux-5.4.0-42-generic-x86_64-with-debian-bullseye-sid
Supports: http, https, ssh, webdav, webdavs
Cache types: hardlink, symlink
Repo: dvc, git
```

**Additional Information (if any):**

If applicable, please also provide a `--verbose` output of the command, eg: `dvc add --verbose`.

### PR DESCRIPTION
**Pull request by daavoo**
Closes #4428

### CODE REVIEW

**Comment by daavoo** [comment]
Thanks @daavoo! It's been a very long time since I looked into this issue, and I'm no longer sure about it, and I'm hesitant to put it in a minor version release.

Let's say I'm using experiments, and my training stage outputs both a model that's cached in dvc and metrics that aren't. With the recent name changes, we made it possible to have duplicate experiments, but it didn't seem so bad because duplicate stages would still be cached and skipped. With this PR, I will end up duplicating all the computation of this training stage.

It also makes dvc much more reliant on dvc.lock. Before this PR, It's possible to have a workflow where you treat dvc.lock as ephemeral and don't commit or share it. As long as you share the run-cache, it's still easy to recover everything. After this PR, there's no way to recover the outputs from a stage with `cache: false` outputs unless you have dvc.lock.

Should dvc be saving the git hash for objects that have `cache: false` instead of the dvc md5? In the typical case for `cache: false` where those files are being tracked by git, it seems like it's still useful to have the run-cache but there's no need to add those files to the dvc cache since they can be found in git objects.

What do you think?

_Originally posted by @dberenbaum in https://github.com/iterative/dvc/issues/8718#issuecomment-1363263137_

**Comment by daavoo** [comment]
> Let's say I'm using experiments, and my training stage outputs both a model that's cached in dvc and metrics that aren't. With the recent name changes, we made it possible to have duplicate experiments, but it didn't seem so bad because duplicate stages would still be cached and skipped. With this PR, I will end up duplicating all the computation of this training stage.

Feels like this goes maybe back to some older comments in https://github.com/iterative/dvc/issues/4428 to add an option to explicitly control `run-cache`?

> It also makes dvc much more reliant on dvc.lock. Before this PR, It's possible to have a workflow where you treat dvc.lock as ephemeral and don't commit or share it. As long as you share the run-cache, it's still easy to recover everything. After this PR, there's no way to recover the outputs from a stage with cache: false outputs unless you have dvc.lock.

The whole `run-cache` behavior is widely undocumented so I am not sure how common this ephemeral `dvc.lock` workflow would be.
 
Still feels that the current `cache: false` + `run-cache` behavior is mostly unexpected for users in most common workflows, perhaps as mentioned above we could disable when `cache: false` but introduce an option to explicitly enable it?

**Comment by dberenbaum** [comment]
> Still feels that the current `cache: false` + `run-cache` behavior is mostly unexpected for users in most common workflows

Yup, I agree. The behavior in this PR (which I haven't tested) seems more expected from what's documented. My concerns are because:

1. It's a complex issue that I hadn't looked at in a long time. Having looked at it again now, I'm less concerned.
2. It will cause noticeable changes in a minor version. Since we are probably moving closer to documented behavior, it's not so bad, but it still may be surprising.
3. It still feels like it's not ideal for some of the scenarios mentioned above. We could keep discussing options to improve this.

We could always do it knowing that we may have to revert until 3.0 if users start to complain.

**Comment by dberenbaum** [comment]
> We could always do it knowing that we may have to revert until 3.0 if users start to complain.

@daavoo Let's go ahead with this PR and keep an eye on whether it causes issues for users.

**Comment by daavoo** [comment]
> @daavoo Let's go ahead with this PR and keep an eye on whether it causes issues for users.

I was thinking that we could try to cover the scenario where only `metrics` or `plots` are `cache: false` by only applying the check to actual `outs` so:

> Let's say I'm using experiments, and my training stage outputs both a model that's cached in dvc and metrics that aren't. 

Would actually create the run-cache.

WDYT @dberenbaum ?

**Comment by dberenbaum** [comment]
Yes, I like it! Let's go with that approach 🙏 .

**Comment by daavoo** [comment]
> Yes, I like it! Let's go with that approach 🙏 .

Updated the P.R.

> Edit: one thing to consider is this doesn't help for dvclive experiments or anyone using top-level metrics, plots, etc. (listing them as regular outputs with cache false in the stage but listing them separately in other top-level sections), but I guess we can revisit that issue as a follow-up.

Need to properly discuss but in that specific scenario(top-level metrics/plots listed as outs with `cache: false`), is there a real need for those files to be listed in `outs`? 
Given that they will be tracked with Git, I would expect that referencing them only in the top-level metrics/plots would be enough

**Review by dberenbaum — APPROVED** [review]
LGTM! 

> Edit: one thing to consider is this doesn't help for dvclive experiments or anyone using top-level metrics, plots, etc. (listing them as regular outputs with cache false in the stage but listing them separately in other top-level sections), but I guess we can revisit that issue as a follow-up.

I was thinking they wouldn't be collected during metrics/plots/params/exp commands, but that's not impacted here, so I don't think it's worth worrying about it.

File: dvc/stage/__init__.py [hunk 1]
```diff
@@ -591,7 +591,15 @@ def run(
...
 
-            self.save(allow_missing=allow_missing, run_cache=not no_commit)
+            no_cache_outs = any(
+                not out.use_cache
+                for out in self.outs
+                if not (out.is_metric or out.is_plot)
+            )
+            self.save(
+                allow_missing=allow_missing,
+                run_cache=not no_commit and not no_cache_outs,
```
Comment (karajan1001) [inline dvc/stage/__init__.py:601]: How about
```python
           cache_outs = not (no_commit or any(
                not out.use_cache
                for out in self.outs
                if not (out.is_metric or out.is_plot)
            ))
```

**Review by karajan1001 — APPROVED** [review]
LGTM,

**Pull request merged**